drivers/pwm/pwm_pca9685.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2022 Nick Ward <nix.ward@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_pca9685
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/pwm.h>
+#include <zephyr/drivers/pwm/pca9685.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pwm_pca9685, CONFIG_PWM_LOG_LEVEL);
+
+#define ADDR_MODE1 0x00
+#define SLEEP	   BIT(4)
+#define AUTO_INC   BIT(5)
+#define RESTART	   BIT(7)
+
+#define ADDR_MODE2	  0x01
+#define OUTDRV_TOTEM_POLE BIT(2)
+#define OCH_ON_ACK	  BIT(3)
+
+#define ADDR_LED0_ON_L	 0x06
+#define ADDR_LED_ON_L(n) (ADDR_LED0_ON_L + (4 * n))
+#define LED_FULL_ON	 BIT(4)
+#define LED_FULL_OFF	 BIT(4)
+#define BYTE_N(word, n)	 ((word >> (8 * n)) & 0xFF)
+
+#define ADDR_PRE_SCALE 0xFE
+
+struct pca9685_config {
+	struct i2c_dt_spec i2c;
+	bool outdrv_open_drain;
+	bool och_on_ack;
+	bool invrt;
+};
+
+struct pca9685_data {
+	uint8_t pre_scale;
+};
+
+static int set_reg(const struct device *dev, uint8_t addr, uint8_t value)
+{
+	const struct pca9685_config *config = dev->config;
+	uint8_t buf[] = {addr, value};
+	int ret;
+
+	ret = i2c_write_dt(&config->i2c, buf, sizeof(buf));
+	if (ret != 0) {
+		LOG_ERR("I2C write [0x%02X]=0x%02X: %d", addr, value, ret);
+	} else {
+		LOG_DBG("[0x%02X]=0x%02X", addr, value);
+	}
+	return ret;
+}
+
+static uint8_t get_reg(const struct device *dev, uint8_t addr)
+{
+	const struct pca9685_config *config = dev->config;
+	uint8_t value;
+	int ret;
+
+	ret = i2c_write_read_dt(&config->i2c, &addr, sizeof(addr), &value, sizeof(value));
+	if (ret != 0) {
+		LOG_ERR("I2C write [0x%02X]=0x%02X: %d", addr, value, ret);
+	}
+	return value;
+}
+
+static void set_pre_scale(const struct device *dev, uint8_t value)
+{
+	uint8_t mode1;
+
+	/* Unblock write to PRE_SCALE by setting SLEEP bit to logic 1 */
+	set_reg(dev, ADDR_MODE1, AUTO_INC | SLEEP);
+
+	mode1 = get_reg(dev, ADDR_MODE1);
+	if ((mode1 & RESTART) == 0x00) {
+		LOG_ERR("RESTART bit should be set");
+	}
+
+	set_reg(dev, ADDR_PRE_SCALE, value);
+
+	/* Clear SLEEP bit - See datasheet section 7.3.1.1 */
+	set_reg(dev, ADDR_MODE1, AUTO_INC);
+	k_sleep(K_USEC(500));
+
+	/* Restart */
+	set_reg(dev, ADDR_MODE1, AUTO_INC | RESTART);
+}
+
+static int pca9685_set_cycles(const struct device *dev, uint32_t pwm, uint32_t period_count,
+			      uint32_t pulse_count, pwm_flags_t flags)
+{
+	const struct pca9685_config *config = dev->config;
+	struct pca9685_data *data = dev->data;
+	uint8_t buf[] = {0x00, 0x00, 0x00, 0x00, 0x00};
+	uint32_t led_off_count;
+	int32_t pre_scale;
+
+	if (pwm >= PCA9685_CHANNEL_CNT) {
```

Comment:
please use `channel` naming, `pwm` is no longer used.

Reply:
Updated